### PR TITLE
feat/caseInsensitiveAutoComplete

### PR DESCRIPTION
**Pull request by rteshnizi**
* Add option `caseInsensitiveAutoComplete` to terminal to perform case insensitive auto complete (to emulate Windows cmd behavior). Default is false.
* Add `simple-demo.html`
* Fix typos in `CONTRIBUTING.md`

### CODE REVIEW

**Comment by jcubic** [comment]
This don't work, when you type lower case and press TAB you get mixed case.

I wanted to add this feature but didn't know how to resolve issue with mixed case words so I've dropped, but his can be resolved with check the case of the last character and if it's lower case, the completion should also be lowercase.

Based on this "[How can I test if a letter in a string is uppercase or lowercase using JavaScript?](https://stackoverflow.com/questions/1027224/how-can-i-test-if-a-letter-in-a-string-is-uppercase-or-lowercase-using-javascrip)" you can use this function:
```javascript
function is_lower(chr) {
  if (!isNaN(chr * 1)) {
     return false;
  } else {
      return chr == chr.toLowerCase();
  }
}
```

Also the option should be `caseInsensitiveAutocomplete`, autocomplete as single work the same as `wordAutocomplete` and it should be taken from top interpreter, if set, like this (same as completion so you can use overwrite the option from push):

```javascript
var top = interpreters.top(), completion, caseSensitive;
if (typeof top.caseInsensitiveAutocomplete != 'undefined') {
   caseSensitive = !top.caseInsensitiveAutocomplete;
} else {
   caseSensitive = !settings.caseInsensitiveAutocomplete;
}
```

And thanks for your contribution.

**Comment by jcubic** [comment]
And one more thing, please fix eslint errors, lines longer then 90 characters, because it make travis build fail. you can check the lines from [this link](https://travis-ci.org/jcubic/jquery.terminal/builds/238942836) or by running `make lint`

**Comment by rteshnizi** [comment]
Thank you for taking a look. I will have to look at it again and fix the eslint error and the name of the option.

I also notice what you are mentioning before submitting the PR and decided to leave it as is. However, I don't think using `is_lower` is the right approach here. A function to guess what is the casing of the string would simply fail in mixed-case strings. Though I agree that it covers some of the cases, I believe fixing the casing of the result shouldn't be an issue since the user is purposefully using `caseInsensitiveAutocomplete` option.

I will see what is the best way to get around the mixed cased issue.

**Comment by jcubic** [comment]
Actually I don't care about mixed case but if text is all lowercase it should be lowercase and if all is uppercase it should be uppercase the else clause can use completion as is.

**Comment by rteshnizi** [comment]
@jcubic I changed the complete() function so when there is only one matched word it replaces the command string with the matched one. How do you feel about this work around?

**Comment by jcubic** [comment]
1. Your code break the tests because it replace the whole command and it will not work when you have `foo BA|` if complation is `'bar'` it will replace the command as `bar` not `foo BAR`, the same happen if cursor is in the middle like `foo BA|zooka` it should complete as `foo BAR|zooka` this is how completion work right now (except the case) so you can't use set_command (if you run `npm install` you can run `make test` to run the tests for command line, you can see how it should work without the caseSensitive option when you open spec file).

2. The workaround would be fine for mixed case but if all character is uppercase I think it would be better to have it all uppercase the same with lowercase. so if you have `foo BA` you get `foo BAR` and if you have `foo ba`  you get `foo bar`, if you have `foo Ba` or `foo bA` and if the completion is 'bar' it could be `foo bar` for both but I think it would be event better to have `foo Bar` and `foo bAr` if the completion is 'bar' and `foo BaR` and `foo bAR` if the completion is `BAR`.

some thing like:

```javascript

                var matched = [];
                for (var i = commands.length; i--;) {
                    if (regex.test(commands[i])) {
                        var match = commands[i];
                        if (quote === '"') {
                            match = match.replace(/"/g, '\\"');
                        }
                        if (!quote && options.escape) {
                            match = match.replace(/(["'() ])/g, '\\$1');
                        }
                        // edit start
                        if (string.toLowerCase() === string) {
                            match = match.toLowerCase();
                        } else if (string.toUpperCase() === string) {
                            match = match.toUpperCase();
                        }
                        // edit end
                        matched.push(match);
                    }
                }
```
that way it will also work for case when you have completion `['barbakan', 'barabazo']` and you try to complete text `B|` you will get BAR| and when you press tab twice you will get echo `BARBAKAN BARABAZO` and you have `b|` will get lowercase.

3. I've just notice this now, the option should be `caseSensitiveAutoComplete` that will be default to true no need for negation.

**Comment by rteshnizi** [comment]
Yep. I noticed the test cases too. I am fixing them slowly. And will fix the option name. Thanks for taking a look.

File: js/jquery.terminal-1.5.0.js [hunk 1]
```diff
@@ -4858,7 +4872,12 @@
                 if (options.escape) {
                     safe = safe.replace(/\\(["'() ])/g, '\\?$1');
                 }
-                var regex = new RegExp('^' + safe);
+                var regex;
+                if (options.caseSensitive) {
+                    regex = new RegExp('^' + safe);
+                } else {
+                    regex = new RegExp('^' + safe, 'i');
```
Comment (jcubic) [inline js/jquery.terminal-1.5.0.js]: it's shorter to use `var regex = new RegExp('^' + safe, options.caseSensitive ? '' : 'i');`

**Comment by rteshnizi** [comment]
I am closing this PR until I have a complete solution.

**Comment by rteshnizi** [comment]
> that way it will also work for case when you have completion ['barbakan', 'barabazo'] and you try to complete text B| you will get BAR| and when you press tab twice you will get echo BARBAKAN BARABAZO and you have b| will get lowercase.

I just want to bring one edge case to your attention using that solution: let's say auto complete options are `['yeah', 'yesSir']`. if user types in `ye` and then press TAB, terminal will echo options as `yeah  yessir`.

**Comment by jcubic** [comment]
You're right so it need to check if the option is on, I think that when it's on (caseSensitive == false), it can work with all lower/all upper and edge case like mixed case completion can be ignored.

**Comment by jcubic** [comment]
You're right so it need to check if the option is on, I think that when it's on (caseSensitive == false), it can work with all lower/all upper and edge case like mixed case completion can be ignored.

But it would be nice if it work like: `yeah yesSir` but it will be more completed, because you need to check if complete on the list of completions is in the same case. if it's in mixed case then it can complete as is like this:
```javascript
                        function same_case(srtring) {
                           var array = string.split('');
                           function all(fn) {
                               return array.filter(function(item) { return item[fn](); }).length;
                           }
                           return all('toLowerCase') || all('toUpperCase');
                        }

                        if (options.caseSensitive && same_case(match)) {
                            if (string.toLowerCase() === string) {
                                match = match.toLowerCase();
                            } else if (string.toUpperCase() === string) {
                                match = match.toUpperCase();
                            }
                        }
```

**Comment by jcubic** [comment]
I've added the feature you wanted in devel branch, you  can check commit 959ae9a7313b91527ac8967bb263d56ec0057645 and if you want you can test and report bugs to this issue #332

**Pull request closed without merging**